lucene/core/src/java/org/apache/lucene/search/FieldExistsQuery.java

```diff
@@ -0,0 +1,165 @@
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.FieldInfos;
+import org.apache.lucene.index.IndexOptions;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+
+/**
+ * A {@link Query} that matches documents that contain either a {@link
+ * org.apache.lucene.document.KnnVectorField}, or a field that indexes norms or doc values.
+ */
+public class FieldExistsQuery extends Query {
+  private String field;
+
+  /** Create a query that will match that have a value for the given {@code field}. */
+  public FieldExistsQuery(String field) {
+    this.field = Objects.requireNonNull(field);
+  }
+
+  public String getField() {
+    return field;
+  }
+
+  @Override
+  public String toString(String field) {
+    return "FieldExistsQuery [field=" + this.field + "]";
+  }
+
+  @Override
+  public void visit(QueryVisitor visitor) {
+    if (visitor.acceptField(field)) {
+      visitor.visitLeaf(this);
+    }
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    return sameClassAs(other) && field.equals(((FieldExistsQuery) other).field);
+  }
+
+  @Override
+  public int hashCode() {
+    final int prime = 31;
+    int hash = classHash();
+    hash = prime * hash + field.hashCode();
+    return hash;
+  }
+
+  @Override
+  public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) {
+    return new ConstantScoreWeight(this, boost) {
+      @Override
+      public Scorer scorer(LeafReaderContext context) throws IOException {
+        FieldInfos fieldInfos = context.reader().getFieldInfos();
+        FieldInfo fieldInfo = fieldInfos.fieldInfo(field);
+        DocIdSetIterator iterator = null;
+
+        if (fieldInfo == null) {
+          return null;
+        }
+
+        if (fieldInfo.hasNorms()) { // the field indexes norms
+          iterator = context.reader().getNormValues(field);
+        } else if (fieldInfo.getVectorDimension() != 0) { // the field indexes vectors
+          iterator = context.reader().getVectorValues(field);
+        } else if (fieldInfo.getDocValuesType()
+            != DocValuesType.NONE) { // the field indexes doc values
+          switch (fieldInfo.getDocValuesType()) {
+            case NUMERIC:
+              iterator = context.reader().getNumericDocValues(field);
+              break;
+            case BINARY:
+              iterator = context.reader().getBinaryDocValues(field);
+              break;
+            case SORTED:
+              iterator = context.reader().getSortedDocValues(field);
+              break;
+            case SORTED_NUMERIC:
+              iterator = context.reader().getSortedNumericDocValues(field);
+              break;
+            case SORTED_SET:
+              iterator = context.reader().getSortedSetDocValues(field);
+              break;
+            case NONE:
+            default:
+              throw new AssertionError();
+          }
+        }
```

Comment:
Maybe add an `else` condition that throws an exception. This query is about finding documents that have a value for a field. If the field exists but doesn't index any of the data structures that can help figure out whether it actually exists, then it means users haven't indexed data correctly, or they are using this query while they shouldn't?

```suggestion
        } else {
          throw new IllegalStateException("FieldExistsQuery requires that the field indexes doc values, norms or vectors, but field '" + fieldInfo.name + "' exists and indexes neither of these data structures");
        }
```

Reply:
Makes sense! I've implemented it in https://github.com/apache/lucene/pull/767/commits/97344ef1726051bbc70869861ecd5d33b9c6ca82.